File: src/main/java/fi/metatavu/oioi/cm/persistence/dao/CustomerDAO.java
package fi.metatavu.oioi.cm.persistence.dao;

import java.util.List;
import java.util.Set;
import java.util.UUID;

import javax.enterprise.context.ApplicationScoped;
import javax.persistence.EntityManager;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;

import fi.metatavu.oioi.cm.persistence.model.Customer;
import fi.metatavu.oioi.cm.persistence.model.Customer_;

/**
 * DAO class for Customer
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class CustomerDAO extends AbstractDAO<Customer> {

  /**
   * Creates new Customer
   * 
   * @param id id
   * @param imageUrl imageUrl
   * @param name name
   * @param creatorId creator's id
   * @param lastModifierId last modifier's id
   * @return created customer
   */
  public Customer create(UUID id, String imageUrl, String name, UUID creatorId, UUID lastModifierId) {
    Customer customer = new Customer();
    customer.setImageUrl(imageUrl);
    customer.setName(name);
    customer.setId(id);
    customer.setCreatorId(creatorId);
    customer.setLastModifierId(lastModifierId);
    return persist(customer);
  }

  /**
   * Updates imageUrl
   *
   * @param imageUrl imageUrl
   * @param lastModifierId last modifier's id
   * @return updated customer
   */
  public Customer updateImageUrl(Customer customer, String imageUrl, UUID lastModifierId) {
    customer.setLastModifierId(lastModifierId);
    customer.setImageUrl(imageUrl);
    return persist(customer);
  }

  /**
   * Updates name
   *
   * @param name name
   * @param lastModifierId last modifier's id
   * @return updated customer
   */
  public Customer updateName(Customer customer, String name, UUID lastModifierId) {
    customer.setLastModifierId(lastModifierId);
    customer.setName(name);
    return persist(customer);
  }

  /**
   * Lists customers by name in set of names
   * 
   * @param names names to list customers by
   * @return list of customers
   */
  public List<Customer> listCustomersByNameIn(Set<String> names) {
    EntityManager entityManager = getEntityManager();
    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
    CriteriaQuery<Customer> criteria = criteriaBuilder.createQuery(Customer.class);
    Root<Customer> root = criteria.from(Customer.class);
    criteria.select(root);
    criteria.where(root.get(Customer_.name).in(names));
    return entityManager.createQuery(criteria).getResultList();
  }

}

File: src/main/kotlin/fi/metatavu/oioi/cm/persistence/dao/CustomerDAO.kt
package fi.metatavu.oioi.cm.persistence.dao

import java.util.UUID
import fi.metatavu.oioi.cm.persistence.model.*
import javax.enterprise.context.ApplicationScoped

/**
 * DAO class for Customer
 *
 * @author Antti Leppä
 */
@ApplicationScoped
class CustomerDAO : AbstractDAO<Customer>() {

    /**
     * Creates new Customer
     *
     * @param id id
     * @param imageUrl imageUrl
     * @param name name
     * @param creatorId creator's id
     * @param lastModifierId last modifier's id
     * @return created customer
     */
    fun create(id: UUID?, imageUrl: String?, name: String?, creatorId: UUID?, lastModifierId: UUID?): Customer? {
        val customer = Customer()
        customer.imageUrl = imageUrl
        customer.name = name
        customer.id = id
        customer.creatorId = creatorId
        customer.lastModifierId = lastModifierId
        return persist(customer)
    }

    /**
     * Updates imageUrl
     *
     * @param imageUrl imageUrl
     * @param lastModifierId last modifier's id
     * @return updated customer
     */
    fun updateImageUrl(customer: Customer, imageUrl: String?, lastModifierId: UUID?): Customer? {
        customer.lastModifierId = lastModifierId
        customer.imageUrl = imageUrl
        return persist(customer)
    }

    /**
     * Updates name
     *
     * @param name name
     * @param lastModifierId last modifier's id
     * @return updated customer
     */
    fun updateName(customer: Customer, name: String?, lastModifierId: UUID?): Customer? {
        customer.lastModifierId = lastModifierId
        customer.name = name
        return persist(customer)
    }

    /**
     * Lists customers by name in set of names
     *
     * @param names names to list customers by
     * @return list of customers
     */
    fun listCustomersByNameIn(names: Set<String?>?): List<Customer> {
        if (names == null || names.isEmpty()) {
            return emptyList()
        }
        val entityManager = entityManager
        val criteriaBuilder = entityManager.criteriaBuilder
        val criteria = criteriaBuilder.createQuery(
            Customer::class.java
        )
        val root = criteria.from(
            Customer::class.java
        )
        criteria.select(root)
        criteria.where(root.get(Customer_.name).`in`(names))
        return entityManager.createQuery(criteria).resultList
    }
}